der/derive/src/sequence.rs

```diff
@@ -50,27 +43,32 @@ impl DeriveSequence {
 
         Self {
             ident: input.ident,
-            lifetime,
+            generics: input.generics.clone(),
             fields,
         }
     }
 
     /// Lower the derived output into a [`TokenStream`].
     pub fn to_tokens(&self) -> TokenStream {
         let ident = &self.ident;
-
-        let lifetime = match self.lifetime {
-            Some(ref lifetime) => quote!(#lifetime),
-            None => default_lifetime(),
+        let mut generics = self.generics.clone();
+
+        // Use the first lifetime parameter as lifetime for Decode/Encode lifetime
+        // if none found, add one.
+        let lifetime: Lifetime = if let Some(lt) = generics.lifetimes().next() {
+            lt.lifetime.clone()
+        } else {
+            let lifetime = default_lifetime();
+            generics.params.insert(
+                0,
+                GenericParam::Lifetime(LifetimeParam::new(lifetime.clone())),
+            );
+            lifetime
         };
 
-        // Lifetime parameters
-        // TODO(tarcieri): support multiple lifetimes
-        let lt_params = self
-            .lifetime
-            .as_ref()
-            .map(|_| lifetime.clone())
-            .unwrap_or_default();
+        // We may or may not have inserted a lifetime.
+        let (_impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+        let (impl_generics, _ty_generics, _where_clause) = generics.split_for_impl();
```

Comment:
This seems a little confusing

Reply:
agreed, but `impl_generics` gets a reference to the `Generics` and the `ToTokens` just consumes the content of the `Generics` to build the token stream.
If I added a lifetime to the generics, then I need `ty_generics` and `where_clause` not to use it, but I still need the generic parameter definition.


```
impl <'der_lifetime, T> DecodeValue<'der_lifetime> for Foo<T>
     \ impl_generics  /                                   \ ty_generics /
```